VisualPinball.Unity/VisualPinball.Unity.Editor/Editors/DragPointsEditor.cs

```diff
@@ -2,72 +2,465 @@
+							cpoint.ScrPos = Handles.matrix.MultiplyPoint(cpoint.WorldPos);
+							if (cpoint.IsSelected)
+							{
+								if (cpoint.DragPoint.IsLocked)
+									cpoint.IsSelected = false;
+								else
+									_selectedCP.Add(cpoint);
+							}
+							HandleUtility.AddControl(cpoint.ControlId, HandleUtility.DistanceToCircle(cpoint.ScrPos, HandleUtility.GetHandleSize(cpoint.WorldPos) * ControlPoint.ScreenRadius * 0.5f));
+						}
+
+						//Setup PositionHandle if some control points are selected
+						if (_selectedCP.Count > 0)
+						{
+							_positionHandlePosition = Vector3.zero;
+							foreach (var sCp in _selectedCP)
+							{
+								_positionHandlePosition += sCp.WorldPos;
+							}
+							_positionHandlePosition /= _selectedCP.Count;
+						}
+
+						if (_curveTravellerVisible)
+							HandleUtility.AddControl(_curveTravellerControlId, HandleUtility.DistanceToCircle(Handles.matrix.MultiplyPoint(_curveTravellerPosition), HandleUtility.GetHandleSize(_curveTravellerPosition) * ControlPoint.ScreenRadius * CurveTravellerSizeRatio * 0.5f));
+					}
+					break;
+
+				case EventType.MouseDown:
+					{
+						if (Event.current.button == 0)
+						{
+							var nearCP = _controlPoints.Find(cp => cp.ControlId == HandleUtility.nearestControl);
+							if (nearCP != null && !nearCP.DragPoint.IsLocked)
+							{
+								bool oldSelection = nearCP.IsSelected;
+								if (!Event.current.control)
+								{
+									ClearAllSelection();
+								}
+								nearCP.IsSelected = oldSelection;
+								nearCP.IsSelected = !nearCP.IsSelected;
+								Event.current.Use();
+							}
+						}
+						else if (Event.current.button == 1)
+						{
+							var nearCP = _controlPoints.Find(cp => cp.ControlId == HandleUtility.nearestControl);
+							if (nearCP != null)
+							{
+								MenuCommand command = new MenuCommand(_itemInspector, nearCP.ControlId);
+								EditorUtility.DisplayPopupMenu(new Rect(Event.current.mousePosition.x, Event.current.mousePosition.y, 0, 0), MenuItems.CONTROLPOINTS_MENUPATH, command);
+								Event.current.Use();
+							}
+							else if (HandleUtility.nearestControl == _curveTravellerControlId)
+							{
+								MenuCommand command = new MenuCommand(_itemInspector, 0);
+								EditorUtility.DisplayPopupMenu(new Rect(Event.current.mousePosition.x, Event.current.mousePosition.y, 0, 0), MenuItems.CURVETRAVELLER_MENUPATH, command);
+								Event.current.Use();
+							}
+						}
+					}
+					break;
+
+				case EventType.Repaint:
+					{
+						if (_needMeshRebuilt)
+						{
+							//Set Meshdirty to true there so it'll trigger again after Undo
+							List<Object> recordObjs = new List<Object>();
+							if (editable != null)
+							{
+								editable.MeshDirty = true;
+								recordObjs.Add(_itemInspector);
+							}
+							recordObjs.Add(_target as Behaviour);
+							_needMeshRebuilt = false;
+							Undo.RecordObjects(recordObjs.ToArray(), _pendingUndoMessage);

```

Comment:
Undo isn't working for me, I wonder if there's something about delaying the record until repaint that unity doesn't like...
In either case it feels a little weird to rely on repaint for this, can we just record the change and trigger mesh rebuild in the end change check blocks?

Reply:
i'll put it back in the prepare, maybe the fix on the MeshExtensions side is enought for the mesh rebuild error.

Reply:
It fixes the Undo for Dragpoints but still no mesh rebuild after the Undo while i am putting both the ItemInspector and the ItemBehavior in the record.